### DISCUSSION TITLE
The package returns an empty string instead of html code or pdf

### DISCUSSION THREAD

**Discussion by condor-bird** [spatie/browsershot#533] [Q&A]
OS: Ubuntu 18.04.4 LTS
Laravel 7
PHP 7.4.3

Js and php packages installed:

```
npm install puppeteer
composer require spatie/browsershot
```

The following code examples are used:

```
Browsershot::url($url)->save($path);
$pdf = Browsershot::html('<h1>Hello world!!</h1>')->pdf();
dd($pdf);
```

However, in both cases, only the empty string "" is returned instead of the expected html content or pdf content.      

Locally on windows, it all works fine.

**Comment by mikebronner** [comment]
How did you solve this? I'm seeing the same issue happening on Forge.

**Reply by ixperiencenl** [reply]
Make sure you use a recent Node version! I myself use the NVM version manager to use multiple Node versions.

**Reply by mikebronner** [reply]
Another issue I found is that I need to install npm dependencies during deployment on each server, it wasn't enough to bundle them in the repo and pull (as we usually do in deployments)

**Reply by cxammar** [reply]
Currently having this issue and I'm desperate. Any advice ?

**Comment by cxammar** [comment]
Hi, I'm facing the same issue, I've got Node v20 and all dependencies installed. Any ideas ?

**Comment by mikebronner** [comment]
Not sure if this answer applies to everyone's situation, but I was able to get it to print content using the following:

```php
        $storagePath = "customer-histories/customer-{$customer->uuid}.pdf";
        $pdf = Browsershot
            ::html(view("pdfs.customer-history", compact("customer"))->render())
            ->emulateMedia("screen")
            ->showBackground()
            ->noSandbox()
            ->setOption('args', ['--disable-web-security'])
            ->format("Letter")
            ->margins(0.25, 0.25, 0.25, 0.25, "in");
        Storage::disk("s3")
            ->put($storagePath, $pdf->pdf());
```

(I do still have the other issue with having to run `npm install` while deploying.)

**Reply by cxammar** [reply]
Hello Mike, many thanks for the quick reply. 

I do also have to run npm install while deploying, I believe it's a pretty standard practice (I use Github Actions for deployment so it's not a big deal).

On the other hand, I don't see what piece of your code does the trick. This is how I build my pdf and I don't see any major differences

```
        $url = route('presupuesto-operaciones', ['budgetData' => $budgetData->toJson()]);

        $pdfBudget = Browsershot::url($url)
            ->setDelay(500)
            ->format('A4')
            ->showBackground()
            ->noSandbox()
            ->timeout(10)
            ->pdf();
```


What do you think about it ?

**Reply by mikebronner** [reply]
Since you're using a URL, you might try with the options settings, and you also may add the emulateMedia option. I think you'll just have to try playing around. Maybe try increasing the timeout to something crazy just to get to a working point, then refine.

**Comment by cxammar** [comment]
So you made it work just by playing with the options ?

**Reply by mikebronner** [reply]
no, look at the differences in my code, and play with those to see what sticks.

**Reply by cxammar** [reply]
Ok thanks Mike, unfortunately it's not doing the trick for me. 

I also tried running a very generic Browsershot to a file and it's giving me a different error. 

` Browsershot::html('<h1>Hello world!!</h1>')->noSandbox()->save('example.pdf');`

` "For some reason Chrome did not write a file at `example.pdf`.\nCommand\n=======\n[]\nOutput\n======\n",`

Which hints that maybe there's some permissions issue going on?

**Reply by cxammar** [reply]
Mike, 

I figured everything out, you were completely right: 

- Empty response from Browsershot is related to Node version not being correct. I was using NVM for managing node versions and had to install through apt get directly to get it working.

- As you mention, you have to npm install puppeteer everytime you deploy (which it's a completely different issue) 


Finally for anyone reading this thread, the cant write file error is related to Puppeteer not locating the Chrome bin and requires more looking into on each specific case, mine was related to a permissions issue of the .cache folder

@mikebronner thanks for your time and patience !

**Comment by nicholaszuccarelli** [comment]
If anyone else happens to come across this problem with Forge, it is definitely to do with write permissions when attempting to run Chromium.
We used a Laravel Forge recipe which installs all the necessary dependencies such as fonts and NodeJS v20
```
apt-get update

apt-get install -y \
    libfreetype6 \
    libharfbuzz0b \
    ca-certificates \
    fonts-freefont-ttf \
    fonts-noto-color-emoji \
    fonts-wqy-zenhei

apt-get install -y libxkbcommon-x11-0

apt-get install -y libxdamage1

apt-get install -y libgbm-dev

curl -fsSL https://deb.nodesource.com/setup_20.x | bash -
apt-get install -y nodejs
```

And as @cxammar mentioned, our server deployment script will then also run `npm install puppeteer` locally.
These steps have resolved the issue for us.
